_P201_VBAN.ino

```diff
@@ -0,0 +1,577 @@
+      //timer = timerBegin(0, 20, true); // 20 Prescaler => 0.25 microseconds  for 32Khz
+      timerAttachInterrupt(timer, &onTimer, true); // binds the handling function to our timer 
+      timerAlarmWrite(timer, 125, true); //x*8Khz sample rate
+      timerAlarmEnable(timer);
+
+
+      //after the plugin has been initialised successfuly, set success and break
+      log = F("VBAN: Initialized");
+      addLog(LOG_LEVEL_INFO,log);
+      success = true;
+      break;
+
+    }
+
+    case PLUGIN_READ:
+    {
+      //code to be executed to read data
+      //It is executed according to the delay configured on the device configuration page, only once
+
+      //after the plugin has read data successfuly, set success and break
+      success = true;
+      break;
+
+    }
+
+    case PLUGIN_WRITE:
+    {
+      //this case defines code to be executed when the plugin executes an action (command).
+      //Commands can be accessed via rules or via http.
+      //As an example, http://192.168.1.12//control?cmd=dothis
+      //implies that there exists the comamnd "dothis"
+
+      //if (plugin_not_initialised)
+       // break;
+
+      //implement any commands here
+
+       break;
+    }
+
+	case PLUGIN_EXIT:
+	{
+	  //perform cleanup tasks here. For example, free memory
+    free(buffer);
+    buffer= NULL;
+	  break;
+
+	}
+    
+    case PLUGIN_ONCE_A_SECOND:
+    {
+      //code to be executed once a second. Tasks which do not require fast response can be added here
+      UserVar[event->BaseVarIndex] = audio_power;
+      success = true;
+      
+      break;
+    }
+  
+    /*
+    case PLUGIN_TEN_PER_SECOND:
+    {
+      //code to be executed 10 times per second. Tasks which require fast response can be added here
+      //be careful on what is added here. Heavy processing will result in slowing the module down!
+      
+      success = true;
+      break;
+    
+    
+    }
+
+  
+    case PLUGIN_FIFTY_PER_SECOND:
+    {
+
+      success = true;
+      break;
+    }
+    */
+  }   // switch
+  return success;
+
+}     //main plugin function
+
+
+
+void adcHandler(void *param) {
+  while (true) {
+    // Sleep until the ISR gives us something to do, or for 1 second
+    ulTaskNotifyTake(pdFALSE, pdMS_TO_TICKS(1000));  
+    if (WiFiConnected()) 
+    if (transmitNow>=0){
+        uint16_t * audio = (uint16_t *)(buffer+transmitNow);
+
+        P201_setFrameCounter((buffer+transmitNow),frameCounter);
+        P201_sendUDP(UDP_IP, UDP_PORT, (buffer+transmitNow),VBAN_PACKET_SIZE);
+        frameCounter++;
+        transmitNow = -1;
+       // addLog(LOG_LEVEL_INFO,"VBAN: sent");
+       int16_t max=0,min=0xffff;
+       
+       for(int i=0;i<SAMPLES;i++)
+       {
+         int16_t val = audio[i];
+         if (val>min) min=val;
+         if (val<max) max=val;
+       } 
+       audio_power = max-min;
+       
+
+      }
+  
+  }
+}
+
+
+
+int16_t last_val=-1;

```

Comment:
Global variables used in an IRAM_ATTR function call should be declared `volatile`.

Reply:
Yep. was a last minute change doing a poor man low pass filter :)

Reply:
actually it is only used in the ISR so is not needed to be volatile.
it is required to be volatile when ISR change or somebody outside changes it and is expected to react to the change imediatelly. 
Caching the value in a register here is ok as speed is important.

Reply:
I'm not entirely sure myself when it must be volatile and when not.
I can imagine the compiler may make some optimizations.
At least it is a 'global' variable, without a namespace so it is essentially a static defined variable.
I'm not entirely sure if the compiler for the ESPxx considers that the same as volatile declared variables, with respect to in what RAM space it will be allocated.

Reply:
the optimizations the compilers are making is how much tocache the value in a registry.
with volatile it always go to read it from memory each time. when optimized the value can be read into a registry and reused some time before gets refreshed from memory again.

Reply:
As these values are not changed by anybody else but the IRS itself I think is safe to not be volatile and actually an optimization here will help being in a registry and access the memory less.

Reply:
Yep, that's the idea of volatile in C++, but as I understood there is also a non-standard effect of it with regards to what memory area will be used to keep the variable.
In C++ a global variable, which is also not placed in a namespace, can be seen the same as if it was declared volatile.
But what I'm not 100% sure of, is whether the compiler for ESP8266/ESP32 does behave the same for global variables without explicit `volatile` attribute.
For ESPxx we have several areas of RAM and one of them is called IRAM. (`IRAM_ATTR` on ESP32 and `ICACHE_ATTR` (I think) on ESP82xx)
The IRAM will always be present and directly available, but we have very limited space of it (on ESP82xx)
So it does not only have to do with cached values getting out of sync, but also with latency to access it and whether it is accessible at all in an ISR.
A node may crash on it, if it isn't directly accessible.

Reply:
I have added the volatile